### PR TITLE
[charts] Add `baseline` property to the `LineChart` `series`

### PR DESCRIPTION
**Pull request by JCQuintas**
Resolves https://github.com/mui/mui-x/issues/14145

This adds a `baseline` props to the line chart series. It only works on the area chart and defines the baseline the area starts from. 

`baseline: 0`
Screenshot 2024-08-09 at 15 14 31

`baseline: -10`
Screenshot 2024-08-09 at 15 14 38

`baseline: 10`
Screenshot 2024-08-09 at 15 14 43

### CODE REVIEW

File: docs/data/charts/lines/lines.md [hunk 1]
```diff
@@ -175,6 +175,16 @@ Different series could even have different interpolations.
 
 {{"demo": "InterpolationDemoNoSnap.js", "hideToolbar": true}}
 
+### Baseline
+
+The area chart draws a `baseline` on the Y axis `0`.
+This is useful as a base value, but customized visualizations may require a different baseline.
+
+To change it, simply provide the wanted `baseline` number of a series.
+Different series can have different `baselines`.
```
Comment (alexfauquette) [inline docs/data/charts/lines/lines.md:184]: It's already not recommended to have multiple areas on a charts (except if staking them) let's not encourage use to have multiple areas with different baselines 😅
```suggestion
```

File: packages/x-charts/src/LineChart/AreaPlot.tsx [hunk 2]
```diff
@@ -97,6 +98,10 @@ const useAggregatedData = () => {
           .x((d) => xScale(d.x))
           .defined((_, i) => connectNulls || data[i] != null)
           .y0((d) => {
+            if (typeof baseline === 'number') {
+              return yScale(baseline)!;
+            }
+
```
Comment (alexfauquette) [inline packages/x-charts/src/LineChart/AreaPlot.tsx:104]: This only impacts the plotting. Here is an example with values fluctuating between 200-300, and with baseline set to 1000.

Notice that:
- The basline has no impact on the max value of the y-axis
- The yèaxis always start at 0 because it's the assumed baseline

image

Comment (alexfauquette) [reply packages/x-charts/src/LineChart/AreaPlot.tsx:104]: Potential solution

1. When computing the extremum values, you replace when needed the base (`d[0]`) by the user defined baseline

2. Instead of providing a number, user can provide `'top' | 'bottom'`. Then you just need to ignore the `d[0]` by doing
  ```diff
  - isArea && axis.scaleType !== 'log' ? ...
  + isArea && axis.scaleType !== 'log' && series[seriesId].baseline === undefined ? ...
  ```

  and in the commented line return 
  ```diff
  - return yScale(baseline)!;
  - return baseline==='top' ? yScale.range[0] : yScale.range[1];
  ```

https://github.com/mui/mui-x/blob/38184fe912eb201d29c8e31b8b11f0f0fd854c51/packages/x-charts/src/LineChart/extremums.ts#L43-L44

**Review by alexfauquette — COMMENTED** [review]
Just one edge case. I think the second option will be better when used with the zoom feature you're currently working on

File: packages/x-charts/src/LineChart/AreaPlot.tsx (hunk 2 above)
Comment (JCQuintas) [reply packages/x-charts/src/LineChart/AreaPlot.tsx]: It behaves a bit weirdly when `baseline` changes the extremums.

This happens because the extremum getter value is then passed into the `nice()` function, it can have unintended consequences, which means I would also have to check the baseline where that is calculated.

But we already have the axis `min/max` values for that. So I believe we should use the min/max values when the axis extremums need to be changed.

screencapture-localhost-3001-playground-area-chart-baseline-2024-08-12-11_20_10

File: packages/x-charts/src/LineChart/extremums.ts [hunk 3]
```diff
@@ -43,8 +43,9 @@ export const getExtremumY: ExtremumGetter<'line'> = (params) => {
         const { area, stackedData } = series[seriesId];
         const isArea = area !== undefined;
 
+        // Since this series is not used to display an area, we do not consider the base (the d[0]).
         const getValues: GetValuesTypes =
-          isArea && axis.scaleType !== 'log' ? (d) => d : (d) => [d[1], d[1]]; // Since this series is not used to display an area, we do not consider the base (the d[0]).
+          isArea && axis.scaleType !== 'log' ? (d) => d : (d) => [d[1], d[1]];
```
Comment (alexfauquette) [inline packages/x-charts/src/LineChart/extremums.ts:48]:
```suggestion
          isArea && axis.scaleType !== 'log' && typeof series[seriesId].baseline !== 'string' ? (d) => d : (d) => [d[1], d[1]];
```

If baseline is min or max we don't want to use the 0 in the extremum computation

File: packages/x-charts/src/models/seriesType/line.ts [hunk 4]
```diff
@@ -82,6 +82,17 @@ export interface LineSeriesType
    * @default 'none'
    */
   stackOffset?: StackOffsetType;
+  /**
+   * The value of the line at the base of the series area.
+   *
+   * If `baseline` is a number, it will be used as the value.
+   *
+   * - `'min'` the minimum value of the series will be used.
+   * - `'max'` the maximum value of the series will be used.
```
Comment (alexfauquette) [inline packages/x-charts/src/models/seriesType/line.ts:91]: IMO it's more interesting to describe the impact than the internal value modification. And the value usecase seems to be very niche compared to the min/max ones


```suggestion
   * - `'min'` the area will fill the space **under** the line.
   * - `'max'` the area will fill the space **above** the line.
   * - `number` the area will fill the space between this value and the line
```

File: docs/data/charts/lines/lines.md [hunk 5]
```diff
@@ -175,6 +175,20 @@ Different series could even have different interpolations.
 
 {{"demo": "InterpolationDemoNoSnap.js", "hideToolbar": true}}
 
+### Baseline
+
+The area chart draws a `baseline` on the Y axis `0`.
+This is useful as a base value, but customized visualizations may require a different baseline.
+
+To change it, simply provide the wanted `baseline` number of a series.
+It is also possible to provide `min` or `max` for the baseline to automatically follow the minimum or maximum value of the series.
```
Comment (alexfauquette) [inline docs/data/charts/lines/lines.md:184]: To encorage using min/max instead of dealing with putting a large value and using `axis.max`

```suggestion
To get the area filling the space above or below the line, set `baseline` to `"min"` or `"max"`.
It is also possible to provide a value to fix the baseline at the desired position.
```

File: docs/data/charts/lines/AreaBaseline.tsx [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+export default function AreaBaseline() {
+  return (
+    <LineChart
+      xAxis={[{ data: [1, 2, 3, 5, 8, 10] }]}
+      series={[
+        {
+          data: [2, -5.5, 2, -7.5, 1.5, 6],
+          area: true,
+          baseline: -10,
```
Comment (alexfauquette) [inline docs/data/charts/lines/AreaBaseline.tsx:12]:
```suggestion
          baseline:  'min',
```

File: docs/data/charts/lines/lines.md [hunk 7]
```diff
@@ -175,6 +175,20 @@ Different series could even have different interpolations.
...
+
+:::info
+Keep in mind the baseline only affects the area visuals. If you want to change the axis minimum or maximum values, you should use the `min/max` property of the axis.
+:::
```
Comment (alexfauquette) [inline docs/data/charts/lines/lines.md:188]: Not exactly if the baseline means we will ignore `0` for extremum computation

Comment (JCQuintas) [reply docs/data/charts/lines/lines.md:188]: I don't get what you mean here 🤔

File: packages/x-charts/src/LineChart/AreaPlot.tsx (hunk 2 above)
Comment (alexfauquette) [reply packages/x-charts/src/LineChart/AreaPlot.tsx]: > It behaves a bit weirdly when baseline changes the extremums.

TLTR: I don't think it's an issue

--- 

Technically speaking, we have two behavior to define:
- where does the area set the baseline
- should the basline impact the extremums.

## Should the baseline impact the extremums

For stacked data, I can not think of a usecase where the answer would be no. Otherwise you would compare some area wherase the base one would have missing part.

image

For a single series, where the area is more aesthetic, seems it can be ignored. Here foe example the NASDAQ evolution during the day. Nobody care of the space between 0 and 16.500. Only the relative evolution matter. Otherwise you would see nothing.

image

## Where does the area set the baseline

I only see two reasons to modify the default 0:

- aesthetic reasons. But in that case, only min/max makes sense
- There is a specific value to compare to. For example the average over time. But in that case, I don't see a reason to pick a value that is always on top or below the line


So basically the weird case you define should not exist

image

File: docs/data/charts/lines/lines.md (hunk 7 above)
Comment (alexfauquette) [reply docs/data/charts/lines/lines.md:188]: Sorry, I forgot a part of the sentence

If the baseline is defined with min/max, the 0 will be ignored for axis extremums

File: packages/x-charts/src/LineChart/AreaPlot.tsx (hunk 2 above)
Comment (JCQuintas) [reply packages/x-charts/src/LineChart/AreaPlot.tsx]: Ah, but those are just working as expected. The issue that appears when we have the `baseline` affect the extremums is that the extremums will get `niced` after. So if we have a non-round number like below, it will not work properly. Then this starts to get more complex and would clash with a feature that already exists `yAxis.min/max`

Screenshot 2024-08-12 at 13 03 52

File: docs/data/charts/lines/lines.md (hunk 7 above)
Comment (JCQuintas) [reply docs/data/charts/lines/lines.md:188]: Ah, this is an effect of your change, I didn't intend it at first. We can keep to the other discussion and I will update it with the outcomes

**Review by alexfauquette — COMMENTED** [review]
Just one last  comment about the impact on stacked line chart. Since you early return the `y0` function, the behavior with stacked line could be wrong. Just warning users about that can do the job

**Review by alexfauquette — APPROVED** [review]
